src/components/Main/Members/Members.tsx

```diff
@@ -0,0 +1,168 @@
+}
+
+export const Members = () => {
+  // hooks
+  const dispatch = useDispatch();
+
+  // initial selected state
+  const initialSelectedState: MemberStateData = {
+    id: -1,
+    firstName: '',
+    lastName: '',
+    email: '',
+    church: { name: ''},
+    disabled: false,
+    roles: [],
+  }
+  // reducer state
+  const members = useSelector(({members}) => members.members);
+  const selectedUser = useSelector(({members}) => members.selectedUser);
+  const localChurch = useSelector(({members}) => members.localChurch);
+
+  // component state
+  const [selectedRows, setSelectedRows] = useState<number[]>([]);
+  const [searchfield, setSearchField] = useState<string>('');
+  const [isConfirmDialogOpen, setIsConfirmDialogOpen] = useState<boolean>(false);
+  const [isAddUserDialogOpen, setIsAddUserDialogOpen] = useState<boolean>(false);
+  const isSelected = (id: number) => selectedRows.indexOf(id) !== -1;
+
+  useEffect(() => {
+    dispatch(onLoadMembers());
+    dispatch(onLoadUser(initialSelectedState));
+  }, [])
+
+  const handleSelectAllClick = (event: React.ChangeEvent<HTMLInputElement>) => {
+    if (event.target.checked) {
+      const newSelectedRows = members.map((member) => member.id);
+      setSelectedRows(newSelectedRows);
+    } else {
+      setSelectedRows([]);
+    }
+  }
+
+  const handleDeleteMembers = async (shouldDelete: boolean) => {
+    setIsConfirmDialogOpen(false);
+    if (shouldDelete) {
+      dispatch(onDeleteMembers(selectedRows))
+      setSelectedRows([]);
+    }
+  };
+
+  const onCloseAddMemberDialog = (shouldAdd: boolean, firstName: string, lastName: string, email: string, password: string) => {
+    setIsAddUserDialogOpen(false);
+    if (shouldAdd && firstName && lastName && email && password && isValidEmail(email)) {
+      dispatch(onAddMember(firstName, lastName, email, password));
+    }
+  }
+
+  const handleRowClick = (event: React.MouseEvent<unknown>, row: MemberStateData) => {
+    event.stopPropagation();
+    const selectedIndex = selectedRows.indexOf(row.id);
+    console.log(selectedRows, selectedIndex, selectedRows.slice(1));
+    let newSelectedRows: number[] = [];
+    if (event.ctrlKey) {
+      if (selectedIndex === -1) {
+        newSelectedRows = newSelectedRows.concat(selectedRows, row.id);
+      } else if (selectedIndex === 0) {
+        newSelectedRows = newSelectedRows.concat(selectedRows.slice(1));
+      } else if (selectedIndex === selectedRows.length -1) {
+        newSelectedRows = newSelectedRows.concat(selectedRows.slice(0, -1));
+      } else if (selectedIndex > 0) {
+        newSelectedRows = newSelectedRows.concat(
+          selectedRows.slice(0, selectedIndex), 
+          selectedRows.slice(selectedIndex + 1),
+        );
+      }
+    } else {
+      if (selectedIndex === -1)
+        newSelectedRows = [row.id]
+      else
+        newSelectedRows = newSelectedRows.concat(selectedRows.slice(0, selectedIndex), selectedRows.slice(selectedIndex + 1));
+    }
```

Comment:
make the conditions even more obvious. makes it a little quicker to read and debug, x times the number of other people who'll read it in the future. a couple conditions i'm still not sure what they do.  
```suggestion
    let newSelectedRows: number[] = [];
    const doingMultiSelect = event.ctrlKey;
    const isNewIndex = selectedIndex === -1;
    if (doingMultiSelect) {
      if (isNewIndex) {
        newSelectedRows = [...selectedRows, row.id];
      } else if (selectedIndex === 0) {
        newSelectedRows = selectedRows.slice(1); // all but first row
      } else if (selectedIndex === selectedRows.length - 1) {
        newSelectedRows = selectedRows.slice(0, -1); // all but last row
      } else if (selectedIndex > 0) {
        newSelectedRows = [ // exclude selected row?
          ...selectedRows.slice(0, selectedIndex),
          ...selectedRows.slice(selectedIndex + 1),
        ];
      }
    } else {
      if (isNewIndex) {
        newSelectedRows = [row.id];
      } else {
        newSelectedRows = [ // exclude selected row?
          ...selectedRows.slice(0, selectedIndex),
          ...selectedRows.slice(selectedIndex + 1),
        ];
      }
    }
```